core/src/main/scala/org/apache/spark/BarrierCoordinator.scala

```diff
@@ -0,0 +1,199 @@
+    }
+  }
+
+  /**
+   * Clean up the array of [[RpcCallContext]]s that correspond to a barrier sync request from a
+   * stage attempt.
+   */
+  private def cleanupSyncRequests(stageId: Int, stageAttemptId: Int): Unit = {
+    val requests = syncRequestsByStageIdAndAttempt.remove((stageId, stageAttemptId))
+    if (requests != null) {
+      requests.clear()
+    }
+    logInfo(s"Removed all the pending barrier sync requests from Stage $stageId (Attempt " +
+      s"$stageAttemptId).")
+  }
+
+  /**
+   * Get the barrier epoch that correspond to a barrier sync request from a stage attempt.
+   */
+  private def getOrInitBarrierEpoch(stageId: Int, stageAttemptId: Int): AtomicInteger = {
+    val barrierEpoch = barrierEpochByStageIdAndAttempt.putIfAbsent((stageId, stageAttemptId),
+      new AtomicInteger(0))
+    if (barrierEpoch == null) {
+      barrierEpochByStageIdAndAttempt.get((stageId, stageAttemptId))
+    } else {
+      barrierEpoch
+    }
+  }
+
+  /**
+   * Send failure to all the blocking barrier sync requests from a stage attempt with proper
+   * failure message.
+   */
+  private def failAllSyncRequests(
+      syncRequests: ArrayBuffer[RpcCallContext],
+      message: String): Unit = {
+    syncRequests.foreach(_.sendFailure(new SparkException(message)))
+  }
+
+  /**
+   * Finish all the blocking barrier sync requests from a stage attempt successfully if we
+   * have received all the sync requests.
+   */
+  private def maybeFinishAllSyncRequests(
+      syncRequests: ArrayBuffer[RpcCallContext],
+      numTasks: Int): Boolean = {
+    if (syncRequests.size == numTasks) {
+      syncRequests.foreach(_.reply(()))
+      return true
+    }
+
+    false
+  }
+
+
+  override def receiveAndReply(context: RpcCallContext): PartialFunction[Any, Unit] = {
+    case RequestToSync(numTasks, stageId, stageAttemptId, taskAttemptId, barrierEpoch) =>
+      // Check the barrier epoch, to see which barrier() call we are processing.
+      val currentBarrierEpoch = getOrInitBarrierEpoch(stageId, stageAttemptId)
+      logInfo(s"Current barrier epoch for Stage $stageId (Attempt $stageAttemptId) is" +
+        s"$currentBarrierEpoch.")
+      if (currentBarrierEpoch.get() != barrierEpoch) {
+        context.sendFailure(new SparkException(s"The request to sync of Stage $stageId (Attempt " +
+          s"$stageAttemptId) with barrier epoch $barrierEpoch has already finished. Maybe task " +
+          s"$taskAttemptId is not properly killed."))
+      } else {
+        val syncRequests = getOrInitSyncRequests(stageId, stageAttemptId)
+        // If this is the first sync message received for a barrier() call, init a timer to ensure
+        // we may timeout for the sync.
+        if (syncRequests.isEmpty) {
+          timer.schedule(new TimerTask {
+            override def run(): Unit = {
+              // Timeout for current barrier() call, fail all the sync requests.
+              val requests = getOrInitSyncRequests(stageId, stageAttemptId)
```

Comment:
what if all the sync requests finish before timeout? then here we may init the request array again.

Reply:
we should have some tests for the timeout behavior, by setting a very small timeout.

Reply:
Em, how about cancel the TimerTask when sync request finished successfully?

Reply:
yea we should do that, but we also need to consider race like sync request finishes and timer triggers at the same time.

Reply:
We will also remove the internal data on stage completed, so I assume the race condition you mentioned won't cause serious issues, the internal data will after all be removed.